w11arm_esd2iso.1.md
% W11ARM_ESD2ISO(1) w11arm_esd2iso 4.0.3
% Paul Rockwell (@Technogeezer on VMware Fusion forum)
% September 2023

# NAME
w11arm_esd2iso - utility to build Windows 11 ARM ISOs from Microsoft ESD files

# SYNOPSIS
**w11arm_esd2iso** [-v]\
**w11arm_esd2iso** [-v] -r *work-dir*\
**w11arm_esd2iso** [-Vh]

# DESCRIPTION
In the first form, a Windows 11 ARM 22H2 ESD (Electronic 
Software Distribution) file is downloaded from Microsoft,
and converted to ISO. This ISO can be used to install Windows 11 ARM on physical
or virtual machines.
	
The utility will prompt for the type of ISO to produce 
(either Home/Pro or Pro/Enterprise) and the language for the created ISO. A list of
available languages will be provided to assist in making the ISO language selection. 

The generated ISO file is placed in the current working directory. The utility
will display the name of the generated ISO file.

In the second form, an interrupted network transfer of the ESD
from Microsoft is resumed from the point of interruption and the ISO build process
will continue. The *work-dir* argument is required, 
and is expected to be a work directory created from a prior failed run of the utility.

In the third form, the utility will either print out it's version or a synopsis of 
command line options, and immediately exit without performing any ISO build.
	
# OPTIONS

**-h**		
: Print a synopsis of usage and exit

**-r** *work-dir*
: Resume an interrupted ESD download using the information from the work directory
*work-dir*. 
	
**-v**	
: Enable verbose output. Use only when providing debugging output.

**-V**	
: Print the version of the utility and exit.


# FILES
./esd2iso_work.*	
: Work directory for ESD download and image creation, located in the current 
working directory.


# DEPENDENCIES
w11arm_esd2iso requires an Apple Silicon (M1/M2) Macs running macOS 13 (Ventura) or later.

Starting with version 3.0, w11arm_esd2iso no longer requires the installation of 
open-source utilities from Homebrew or MacPorts. 
All required utilities are provided in the distribution zip file.

# EXIT VALUES
**0**
: Successful creation of the ISO file

**Non-zero** 
: Unsuccessful creation of the ISO file
	
# NOTES


There are no intentions to make this script run on Windows or Intel Macs.

w11arm_esd2iso only builds ISO media for Windows 11 ARM 22H2. It does not build media 
for any other versions of Windows or Windows Server.

The current working directory must be changed to the directory where the distribution zip
file of w11arm_esd2iso was extracted. The utility must be executed from that directory 
otherwise the required bundled components will not be found. 

Have at least 12 GB of free disk space to run w11arm_esd2iso. Checks are performed for
sufficient disk space and the utility will refuse to run if space isn't available.
available. 

The w11arm_esd2iso utility contains bundled versions of the open-source wimlib-imagex(1) and 
aria2c(1) utilities. These will be used if the utilities are not found to be installed
with Homebrew/MacPorts. This is done for ease of use - the user 
does not have to go through the complexity of installing Xcode command line tools, 
Homebrew/MacPorts, and the packages containing these utilities. Im the unlikely event that 
existing user-installed versions of aria2c and wimlib-imagex present a problem, 
an uninstall of Homebrew/MacPorts is not necessary. The default behavior of using the 
bundled utilites can be restored by removing the Homebrew/MacPorts binary directory from 
the $PATH variable of the shell being used to
run w11arm_esd2iso. 

The work directory will be deleted upon successful ISO creation and when most 
errors are encountered. It will not be deleted if a transfer of the ESD can be
restarted using the -r option, or if the final phase of the ESD creation
fails.  

The command line syntax has changed starting with V4.0 and is not 100% compatible with 
V3.0.1 and earlier. The functions provided by command line options and arguments are now
provided by interactively prompting the user during the build process. 

The -a option has been removed as aria2c(1) is now used for all downloads. 

w11arm_esd2iso automatically enables retry and restart features that handle the interruption 
of ESD downloads from Microsoft due to network connectivity issues.  The 
utility will automatically retry an interrupted download up to 20 times. After 20 retries,
the utility will exit, and will display a command (containing the -r option) that can be 
copy/pasted to resume the ESD download. 

The -r option can be used as many times as necessary as directed by the utility to 
complete the download. In all honesty, it shouldn't need to be used more than 
once unless you have a really, really bad network connection. Each restart attempt
using the -r option will resume a download from the point of interruption of the
last execution of the utility. 

w11arm_esd2iso will verify the SHA1 hash of the downloaded ESD file against the hash value
provided by Microsoft. If the verification fails (indicating a corrupt
download), the utility will exit with an error message.

More extensive guidance on the use of w11arm_esd2iso to create Windows 11 ARM virtual
machines with VMware Fusion 13 on Apple Silicon can be found in the 
Unofficial Fusion 13 for Apple Silicon Companion Guide which can be found at
https://communities.vmware.com/t5/VMware-Fusion-Documents/The-Unofficial-Fusion-13-for-Apple-Silicon-Companion-Guide/ta-p/2939907

# EXAMPLES

**./w11arm_esd2iso**
: Download and build a Windows 11 ARM ISO. The user will be prompted for both the type
of ISO to create, and the language of the ISO. 

**./w11arm_esd2iso -r ./esd2iso_work.abcdef**
: Restart a interrupted download from the point of interruption using the
work directory ./esd2iso_work.abcdef left behind from a previous run of the utility.

# SEE ALSO
wimlib-imagex(1), aria2c(1), hdiutil(1)

# CREDITS
Information for obtaining Microsoft ESD distributions and
Microsoft Product catalog from b0gdanw "ESD to ISO on macOS.txt" 
https://gist.github.com/b0gdanw/e36ea84828dbd19e03eff6158f1fc77c

# LICENSE
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
	
# SUPPORT
No production level support is available for this script. The author 
provides best effort response to any issues with no service level targets.

Any questions about this utility should be posted to the VMware Fusion Discussions 
forum: https://communities.vmware.com/t5/VMware-Fusion/ct-p/3005-home

If you find an error please run the utility with the -v option, then attach the entire
output of the utility in a zip file to your post.


CHANGELOG
CHANGELOG

Version 4.0.3 (17-September-2023)

	Bug fix: w11arm_esd2iso would fail with error 134 when trying to download the catalog information
	from Microsoft when running under macOS Sonoma. This error was due to an issue
	with the aria2c utility that w11arm_esd2iso uses. This issue was fixed by setting an
	environment variable recommended by the aria2c developers. 

Version 4.0.2 (13-July-2023)

	Version 4.0.2 fixes a bug where installation of an open-source version of 'tar' 
	through Homebrew or MacPorts caused a failure to extract the Windows 11 catalog of
	available downloads. This bug was caused by an open-source version of 'tar' being
	installed by Homebrew and found on the user's PATH variable before the macOS 
	installed version. Thie bug is fixed by invoking the 'tar' utility using its full 
	path instead of relying on it being found through the PATH variable.
	
Version 4.0.1 (03-July-2023)

	Version 4.0.1 adds verification of the SHA1 hash of the ESD download to the hash
	value provided by Microsoft. This will catch rare instances of a
	corrupted download.  

Version 4.0 (28-June-2023)

	Version 4.0 is now an interactive utility that makes it even easier to create the Windows
	11 ARM ISO.  
	
	THe utility now asks the user for the type of ISO to create (either Home/Pro or Pro/Enterprise)
	and asks the user to select the language of the ISO from a list of valid languages
	that it prints. It is no longer required to specify this information on the command line
	or to run another option of the utility to print out the valid Windows languages.

	As part of this change, the command line syntax is not compatible with versions 3.0.1
	and earlier. The following changes have been made to the command line syntax:
 
 	- The esd-language argument on the command line is no longer required and has been removed. 
 	  This is replaced by the interactive dialog within the utility.
 	  
 	- The -l option to list the available languages has been removed.
 	  This is now automatically done as part of the new interactive dialog.
 	  
 	- The -b option to specify the Pro/Enterprise ISO hs been removed.
 	  This is replaced by a question for the user to select the editions in the 
 	  new interactive dialog. 
 	  
	The -r option remains and is still used to resume an interrupted download.
	
	Version 4.0 recognizes if the wimlib or aria2 open-source utilities from Homebrew or 
	MacPorts are available via the users $PATH veriable, and will use them if they are found.
	If they aren't found, the bundled versions will be used. 
	
	The amount of disk required by the ISO build process has been reduced. 
	The downloaded ESD file is deleted immediately before step 4 (the creation of the 
	ISO from the installation image) as it is no longer needed at that point. This saves
	approximately 5 GB of disk space over versions 3.0.1 and earlier during the build 
	process. 
	 
	
Version 3.0.1 (21-May-2023)

	Version 3.0 would display macOS Gatekeeper errors such as "unidentified
	developer" and "unable to verify with Apple" which prevented the utility from running.
	These errors were displayed when the the utility tried to run any of the bundled open source
	components.
	The errors could be worked around, but would require multiple
	manual interventions by the user to allow w11arm_esd2iso to run.
	This has been fixed in 3.0.1, and no manual intevention by the user is required.

Version 3.0 (19-May-2023)	

	Now released under GPLv2 license.

	Removal of manpage documentation from README file.
	
	Creation of manpage documentation in pdf format.
	
	Rename of utility from w11arm_esd2iso.sh to w11arm_esd2iso.

	Installation of aria2 and wimlib from Homebrew or MacPorts is no longer required. The
	necessary utilities that w11arm_esd2iso needs are now bundled with the distribution.

	Tests required for existence of Homebrew or MacPorts have been removed since they
	are no longer necessary. 

	Removal of -a option - aria2c is used for all downloads and restarts are
	automatically enabled.
	
Version 2.1 (22-Mar-2023)

	Additional checks for required utilities. 
	
	Check if $PATH contains the Homebrew or MacPorts directores. Add them if
	they are present and are not on $PATH already.
	
	Examples now appear in the README file

Version 2.0 (15-Mar-2023)

	Enhanced error detection.
	
	Reduction of "chatty" output when the -v (verbose) option is not selected. 
	
	The -v (verbose) option now includes more information for use in debugging
	(and if you're really interested in what goes on under the hood).
	
	Addition of -V option to print the script version and exit.
	
	When using the -a option, the script will now retry an interrupted ESD download 
	up to 20 times before aborting the script. This should reduce the need to use the 
	-r option to restart interrupted ESD downloads due to network issues.
	
	Some internal changes for efficiency.
	

Version 1.0 - Initial release



w11arm_esd2iso.sh
#!/bin/sh
#
#
# Credit: Location and methods of obtaining Microsoft ESD distributions and
# Microsoft Product catalog from b0gdanw "ESD to ISO on macOS.txt" https://gist.github.com/b0gdanw/e36ea84828dbd19e03eff6158f1fc77c
#
# This script is offered as-is, with NO GUARANTEES OR WARRANTIES
#


verbosityLevel=0
aOption=0
bOption=0
hOption=0
lOption=0
rOption=0
langCodes=""

version="w11arm_esd2iso 2.0 (15-Mar-2023)\n"
requires="wiminfo wimapply wimexport"

usage() {
  echo "Usage:\n"
  echo "$0 [-abhvV] esd-language"
  echo "$0 [-vVh] -r work-dir"
  echo "$0 [-vVh] -l\n\nArguments:"
  echo "\tesd-language\\tValid Windows 11 language tag"
  echo "\twork-dir\tWork directory for restarting failed download"
  echo "\nOptions:"
  echo "\t-a\tUse aria2c for ESD download instead of curl"
  echo "\t-b\t  Download business (Pro/Enterprise) ESD instead of consumer (Home/Pro)"
  echo "\t-h\tPrint usage and exit"
  echo "\t-v\tEnable verbose output"
  echo "\t-V\tPrint script version and exit"
  echo "\t-r work-dir\n\t\tRestart failed ESD download from a prior execution using work-dir"
}
printVersion() {
	echo $version
}
verboseOn() {
	if [ $verbosityLevel -eq 0 ]; then
		return 1
	else
		return 0
	fi
}
printLanguages() {
	local langCode
	local langDesc
	
	printf '%s\n\n' "Valid Microsoft ESD language tags are:"
	printf '%-20s %-30s\n' "Language tag"	"Language"
	printf '%-20s %-30s\n'  "------------" "--------------------"
    for langCode in $langCodes; do
      langDesc="$(xpath -q -n -e '//File[LanguageCode="'$langCode'"]/Language' $workingDir/w11arm_products.xml | sed -E -e s/'(<Language>)|(<\/Language>)'//g )"
      printf '%-20s %-30s\n' "$langCode" "$langDesc"
    done
    return 0
}
downloadCatalog() {
	local wDir
	local editionType
	local showMeter
	local retVal
	local edQuery
	local editionName
	
	wDir=$1
	editionType=$2
	showMeter="--no-progress-meter"
	
	if [ ! -d $wDir ]; then
		echo "ERROR: $wDir is expected to exist for downloadCatalog but doesn't"
		echo "Please report this error"
		exit 1
	fi
	
	#
	# Download and process the Windows Product Catalog so we know what languages are available
	# and the URLs for the downloads
	#
	
	if [ $editionType -eq 0 ]; then
		editionName="Consumer (Home/Pro)"
	else
		editionName="Business (Pro/Enterprise)"
	fi
	
	echo "\nPlease wait while information is being downloaded from Microsoft for the $editionName ESDs"
	
	verboseOn && showMeter=""
	curl -L "https://go.microsoft.com/fwlink?linkid=2156292" $showMeter -o $wDir/catalog.cab
	retVal=$?
	if [ $retVal -ne 0 ]; then
		return $retVal
	fi
	( cd $wDir; tar xf catalog.cab products.xml; sed -i .bak 's/\r//' products.xml )

	
	#
	# Get only the W11 Pro ARM64 products
	#
	echo '<Catalog>' > $wDir/w11arm_products.xml
	if [ $editionType -eq 0 ]; then
		edQuery='//File[Architecture="ARM64"][Edition="Professional"]'
	else
		edQuery='//File[Architecture="ARM64"][Edition="Enterprise"]'
	fi
	
	xpath -q -n -e $edQuery $wDir/products.xml >> $wDir/w11arm_products.xml
	
	echo '</Catalog>'>> $wDir/w11arm_products.xml

	#
	# Get the available languages
	# 
	
	langCodes=$(xpath -q -n -e '//File/LanguageCode' $wDir/w11arm_products.xml | sed -E -e s/"<[\/]?LanguageCode>"//g | sort)
	echo "Processing complete"
	return 0
}

setupEsdDownload(){

###################
# Find the desired language variant file 
#
# Now that we know we have an image available in the desired language,
# build the file names for the download URL (from the Microsoft catalog info),
# the file name for the ESD that resides in the working directory
# and the name of the ISO we're going to generate
###################
	
	xpath -q -n -e '//File[LanguageCode="'$esdLang'"]' $workingDir/w11arm_products.xml >$workingDir/esd_edition.xml
	esdURL=$(xpath -n -q -e '//FilePath' $workingDir/esd_edition.xml | sed -E -e s/'(<FilePath>)|(<\/FilePath>)'//g)
	buildName=$(xpath -n -q -e '//FileName' $workingDir/esd_edition.xml | sed -E -e s/'(<FileName>)|(\.esd<[\/]FileName>)'//g)

	esdFile=$workingDir/${buildName}.esd
	isoFile=./${buildName}.iso

	if verboseOn ; then
	    echo "\nesdURL = $esdURL"
	    echo "buildName = $buildName"
	    echo "esdFile = $esdFile"
	    echo "isoFile = $isoFile/n"
	fi

	echo "Selected language of the ISO is $esdLang \c"
	xpath -n -q -e '//Language' $workingDir/esd_edition.xml | sed -E -e s/'(<Language>)|(<\/Language>)'//g
	verboseOn && echo "The ESD will be downloaded from\n$esdURL"
	echo "The generated ISO will be $isoFile"
	return 0
}

downloadEsd() {

	local retVal
	local retryCount
	local retryLimit
	local showProgress
	
	retryLimit=20
	retryCount=0
	
	verboseOn && echo "NOTE: Using $downloadApp for ESD download"
	#
	# Get the ESD from Microsoft
	#
	
	if [ $downloadApp = "curl" ]; then
	   showProgress="--progress-bar"
	   verboseOn && showProgress=""
		curl --keepalive-time 300 $showProgress $esdURL -o $esdFile;
		retVal=$?
	else
		showProgress="--summary-interval=0"
	    verboseOn && showProgress="--summary-interval=120"
	    retVal=1
	    while [ $retVal -ne 0 -a $retryCount -lt $retryLimit ]
		do
			retryCount="$((retryCount + 1))"
			echo "ESD Download attempt $retryCount"
			aria2c $showProgress -d $workingDir --file-allocation=none $esdURL
			retVal=$?
			if [ $retVal -ne 0 ]; then
				echo "Download interrupted, \c"
				if [ $retryCount -eq $retryLimit ]; then
					echo "retry limit reached"
				else
					echo "retrying"
				fi
			fi
		done
	fi
	if [ $retVal -eq 0 ]; then
		verboseOn && wiminfo $esdFile
	fi
	return $retVal
}

extractEsd(){	
	
	local eFile
	local eDir
	local retVal
	local esdImageCount
	local bootWimFile
	local installWimFile
	local images
	local imageIndex
	local imageEdition
	local beQuiet
	
	eFile=$1
	eDir=$2
	beQuiet="--quiet"
	bootWimFile=$eDir/sources/boot.wim
	installWimFile=$eDir/sources/install.wim
	images="4 5"
	
	verboseOn && beQuiet=""
	
	esdImageCount=$(wiminfo $eFile | awk '/Image Count:/ {print $3}')
	verboseOn && echo "image count in ESD: $esdImageCount"

	#
	# Extract image 1 in the ESD to create the boot environment
	#

	echo "\nApplying boot files to the image"
	wimapply $eFile 1 $eDir $beQuiet 2>/dev/null
	retVal=$?
	if [ $retVal -ne 0 ]; then
		echo "ERROR: Extract of boot files failed"
		return $retVal
	fi

	echo "Files applied"

	################
	# Create the boot.wim file that contains WinPE and Windows Setup
	# Images 2 and 3 in the ESD contain these components
	#
	# Important: image 3 in the ESD must be marked as bootable when
	# transferred to boot.wim or else the installer will fail
	################

	echo "\nAdding WinPE and Windows Setup to the image"
	wimexport $eFile 2 $bootWimFile --compress=LZX --chunk-size 32K $beQuiet
	retVal=$?
	if [ $retVal -ne 0 ]; then
		echo "ERROR: Add of WinPE failed"
		return $retVal
	fi

	wimexport $eFile 3 $bootWimFile --compress=LZX --chunk-size 32K --boot $beQuiet
	retVal=$?
	if [ $retVal -ne 0 ]; then
		echo "ERROR: Add of Windows Setup failed"
		return $retVal
	fi
	echo "Files added to image"
	
	verboseOn && wiminfo  $bootWimFile

	echo "\nAdding Windows editions to image"

	################
	# Create the install.wim file that contains the files that Setup will install
	# Images 4, 5, (and 6 if it exists) in the ESD contain these components
	################
	

	[ $esdImageCount -eq 6 ] && images="$images 6"
	
	for imageIndex in $images; do
		imageEdition="$(wiminfo $eFile $imageIndex | grep '^Description:' | sed 's/Description:[ \t]*//')"
		echo "\tAdding $imageEdition to the image"
		wimexport $eFile $imageIndex $installWimFile --compress=LZMS --chunk-size 128K $beQuiet
		retVal=$?
		if [ $retVal -ne 0 ]; then
			echo "ERROR: Addition of $imageIndex to the image failed"
			return $retVal
		fi
	done

	echo "All Windows editions added to image"	
	
	verboseOn && wiminfo $installWimFile

	return 0
}

buildIso(){
	local iDir=$1
	local iFile=$2
	local elToritoBootFile
	
	iDir=$1
	iFile=$2
	
	if [ -e $iFile ]; then
	  echo "\tWARNING: File $iFile exists, removing it"
	  rm -rf $iFile
	fi

	elToritoBootFile=$iDir/efi/microsoft/boot/efisys.bin
	
	#
	# Create the ISO file
	#

	hdiutil makehybrid -o $iFile -iso -udf -hard-disk-boot -eltorito-boot $elToritoBootFile $iDir
	return $?
}
#-------------------
#
# Start of program
#
#-------------------


#-------------------
# 
# Process arguments
# 
#-------------------

while getopts ":abhlr:vV" opt; do
  case $opt in
    a)
    	aOption=1
    	;;
    b)
    	bOption=1
    	;;
    h)
    	usage
    	exit 1
    	;;

    r)
    	rOption=1
    	workingDir=$OPTARG
    	;;
    l)
    	lOption=1
    	;;
	v)
    	verbosityLevel=$((verbosityLevel + 1))
    	;;
	V)
    	printVersion
    	exit 1
    	;;
    :)
    	echo "Option -$OPTARG requires an argument"
    	usage
    	exit 1
    	;;
    
    \?)
    	echo "Invalid option: -$OPTARG\n"
    	usage
    	exit 1
    	;;
    esac
done
shift "$((OPTIND-1))"

#-------------------
#
# Check number of arguments
# One argument is allowed when not using the -r option for restart
# If using the -r or -l options no arguments are allowed
#
#-------------------

if [ $# -gt 1 ]; then
	echo "ERROR: Too many arguments"
	usage
	exit 1
fi

if [ $rOption -eq 0 -a $lOption -eq 0 ]; then
	if [ $# -lt 1 ]; then
		echo "ERROR: esd-language argument is required"
		usage
		exit 1
	fi
fi
if [ $rOption -eq 1 ]; then
	if [ $lOption -eq 1 -o  $aOption -eq 1 ]; then
		echo "ERROR: the -a and -l options can not be used with the -r option"
		usage
		exit 1
	fi
	if [ $# -eq 1 ]; then
		echo "ERROR: too many arguments when using -r option"
		usage
		exit 1
	fi
fi
if [ $lOption -eq 1 ]; then
	if [ $rOption -eq 1  -o  $aOption -eq 1  ]; then
		echo "ERROR: the -a and -r options can not be used with the -l option"
		usage
		exit 1
	fi
	if [ $# -eq 1 ]; then
		echo "ERROR: too many arguments when using -l option"
		usage
		exit 1
	fi
fi

#-------------------
# Check for for wiminfo, wimapply and wimexport 
#
# all of these are from open source package wimlib
# Warn the user if any of them are not here and abort
#-------------------

cmdPath=""
for utilPath in "/opt/homebrew/bin" "/opt/local/bin" "/usr/local/bin"; do
	if [ -d "$utilPath" ]; then
		cmdPath="${utilPath}"
		verboseOn && echo "Utility path is $utilPath"
		break
	fi;
done
if [ "x$cmdPath" = "x" ]; then
	echo "ERROR: Neither Homebrew or MacPorts appear to be installed"
	echo "Please install one of them and then install required packages"
	exit 1
fi

# 
# If for some reason the user has installed Homebrew or MacPorts, but has not
# put the base directory into $PATH, lets do it for them
#

echo $PATH | grep -E "$cmdPath"'((:)|($))'
if [ $? -ne 0 ]; then
	verboseOn && echo "NOTE: $cmdPath exists, but isn't on shell's PATH. Adding it."
	PATH="${cmdPath}:${PATH}"
	verboseOn && echo "Path is now set to $PATH"	
fi

notFound=0
for i in $requires; do
    which -s $i
    if [ $? -ne 0 ]; then
      echo "ERROR: Required program $i not found"
      notFound=1
    fi
done
if [ $notFound -eq 1 ]; then
      echo "Please install required packages from Homebrew or MacPorts"
      exit 1
fi 

#-------------------
#
# If the -a option was specified, see if aria2c exists
# If we are doing a restart and aria2c doesn't exist, error out since a restartable
# download can only occur with aria2c
# If we aren't doing a restart and aria2c doesn't exist, then default back to curl
#
#-------------------

downloadApp="curl"
if [ $aOption -eq 1 ]; then
	which -s aria2c
	if [ $? -eq 0 ]; then
		downloadApp="aria2c"
	else
		if [ $rOption -eq 1 ]; then
			echo "ERROR: Download restart requested, but aria2c can not be found"
			echo "Restart can not be performed"
			exit 2
		else
		    echo "WARNING: aria2c not found, -a option ignored"
		    aOption=0
		fi

   	fi
fi

if [ $rOption -eq 0 -o $lOption -eq 1 ]; then
	################
	#
	# If normal processing or the -l option has been specified:
	#
	# 	Make a temp in the current working directory where we're going to do all of our work
	# 	Download the catalog after the temp directory has been created
	#
	################

	workingDir="$(mktemp -q -d ./esd2iso_temp.XXXXXX)"
	if [ $? -ne 0 ]; then
		echo "Unable to create work directory, exiting"
		exit 1
	fi
	downloadCatalog $workingDir $bOption
	retVal=$?
	if [ ! $retVal ]; then
		echo "ERROR: Catalog download failed with error $retVal"
		rm -rf $workingDir
		exit 1
	fi
fi

if [ $lOption -eq 1 ]; then
	printLanguages
	rm -rf $workingDir
    exit 1
fi 

if [ $rOption -eq 0 ]; then

	################
	#
	# Normal processing, no restart
	#
	################
		
	freeSpace=$(df -g . | awk '/dev.*/ {print $4}')
	verboseOn && echo "Free space is $freeSpace GB"
	if [ $freeSpace -lt 15 ]; then
		echo "ERROR: You have insufficent space on this disk to complete the ISO build."
		echo "This script requires approximately 15 GB of free disk space and you have "
		echo "$freeSpace GB remaining."
		exit 1
	else
		if [ $freeSpace -lt 20 ]; then
			echo "WARNING: This script typically requires 12-15 GB of free disk space to complete"
			echo "You have $freeSpace GB available. You may run out of disk space during the process."
		fi
	fi
    
	################
	#
	# extDir is the "extract directory" where we're going to extract the ESD 
	# and evenutally build the ISO from. It's a subdirectory of the working/temp directory
	#
	################

	extDir=$workingDir/ESD_ISO
	mkdir $extDir

	esdLang=$(echo $1 | tr "[:upper:]" "[:lower:]")

	if [ $esdLang != $1 ]; then
		echo "NOTE: Language tag $1 has been converted to lower case"
	fi

	################
	#
	# Verify that the language is supported by Windows
	#
	################
	
	echo "\nVerifying language $esdLang is available\c"

	found=0
	for i in $langCodes; do
	 if [ $esdLang = $i ]; then 
	   found=1; 
	 fi;
	done

	if [ $found -eq 0 ]; then
	  echo "...FAILED\nERROR: $esdLang is not a valid language tag"
	  printLanguages
	  rm -rf $workingDir
      exit 1
	fi
	
	echo "...verified"

	echo "\nStep 1: Getting information about the requested $esdLang ESD file\n"
	setupEsdDownload
	echo "\nStep 1 complete"

	echo "\nStep 2: Downloading ESD from Microsoft with $downloadApp"
	downloadEsd
	retVal=$?
	if [ $retVal -ne 0 ]; then
		echo  "ERROR: ESD download interrupted with error $retVal"
		if [ $aOption -eq 1 ]; then 
			################
			#
			# Write out values of variables that we'll need for restart
			# in a flag file restartOK that we set in the work directory
			#
			################
			
			touch $workingDir/restartOK
			echo "isoFile $isoFile" >> $workingDir/restartOK
			echo "esdFile $esdFile" >> $workingDir/restartOK
			echo "esdURL $esdURL" >> $workingDir/restartOK
			echo "ESD download can be restarted with the following command:"
			echo "$0 -r $workingDir"
		else
			rm -rf $workingDir
		fi
		exit 1
	fi

else
	################
	#
	# Restart option selected
	#
	################
	
	################
	#
	# See if the work directory exists and if it supports restart
	#
	################
	
	if [ ! -d $workingDir ]; then
		echo "ERROR:Work directory $workingDir does not exist, exiting"
		exit 2
	fi
	if [ ! -f "$workingDir/restartOK" ]; then
		echo "ERROR: The work directory $workingDir does not support restart, exiting"
		exit 2
	fi;
	echo "\nRestarting interrupted download using work directory $workingDir\n"

	################
	#
	# Read variables needed for restart from the restartOK file
	#
	################
	
	extDir=$workingDir/ESD_ISO
	isoFile=$(cat $workingDir/restartOK | awk '/isoFile/ { print $2 }')
	esdFile=$(cat $workingDir/restartOK | awk '/esdFile/ { print $2 }')
	esdURL=$(cat $workingDir/restartOK | awk '/esdURL/ { print $2 }')

	if verboseOn ; then
	    echo "\nesdURL = $esdURL"
	    echo "esdFile = $esdFile"
	    echo "isoFile = $isoFile/n"
	fi

	echo "Restarting Step 2: Downloading ESD from Microsoft with $downloadApp"
	downloadEsd
	retVal=$?
	if [ $retVal -ne 0 ]; then
		echo  "ERROR: ESD download failed with error $retVal"
		echo "ESD download can be restarted with the following command:"
		echo "$0 -r $workingDir"
		exit 1
	else
		rm $workingDir/restartOK
	fi
fi

################
#
# Regardless of whether restart was asked for or not, at this point
# we have an ESD downloaded. 
#
################
	
echo "\nStep 2 complete - ESD downloaded"	
# echo "no need to build anything further -we know that works"
# rm -rf $workingDir
# exit 0
	
echo "\nStep 3: Building installation image from ESD distribution"
extractEsd $esdFile $extDir
retVal=$?
if [ $retVal -ne 0 ]; then
	echo "Installation image build failed with error code $retVal"
	echo "Work directory $workingDir was not deleted, use for debugging"
	exit 1
fi
echo "\nStep 3 complete - installation image built"

echo "\nStep 4: Creating ISO $isoFile from the installation image\n"
if buildIso $extDir $isoFile ; then
    echo "Step 4 complete - ISO created"
else
    echo "ERROR: ISO was NOT created"
    echo "Working directory $workingDir was not deleted, use for debugging"
    exit 1
fi

echo "\nCleaning up work directory"
rm -rf $workingDir
echo "Done!"
exit 0
